src/main/scala/analysis/DSAUtility.scala

```diff
@@ -0,0 +1,983 @@
+import specification.{ExternalFunction, SpecGlobal, SymbolTableEntry}
+
+import scala.util.control.Breaks.{break, breakable}
+import scala.collection.mutable
+
+object NodeCounter {
+  var counter: Int = 0
+
+  def getCounter: Int =
+    counter = counter + 1
+    counter
+}
+
+/**
+ * Data Structure Graph for DSA
+ * @param proc procedure of DSG
+ * @param constProp
+ * @param varToSym mapping flow-sensitive (position sensitive) mapping from registers to their set of symbolic accesses
+ * @param globals
+ * @param globalOffsets
+ * @param externalFunctions
+ * @param reachingDefs
+ * @param writesTo
+ * @param params
+ */
+class DSG(val proc: Procedure,
+          constProp: Map[CFGPosition, Map[Variable, FlatElement[BitVecLiteral]]],
+          varToSym: Map[CFGPosition, Map[Variable, Set[SymbolicAccess]]],
+          globals: Set[SymbolTableEntry], globalOffsets: Map[BigInt, BigInt],
+          externalFunctions: Set[ExternalFunction],
+          val reachingDefs: Map[CFGPosition, Map[Variable, Set[CFGPosition]]],
+          val writesTo: Map[Procedure, Set[Register]],
+          val params: Map[Procedure, Set[Variable]]
+         ) {
+
+  // DSNodes owned by this graph, only updated once analysis is done,
+  val nodes: mutable.Set[DSN] = mutable.Set()
+
+
+  // represent callees in proc
+  val callsites: mutable.Set[CallSite] = mutable.Set()
+
+  val mallocRegister = Register("R0", 64)
+  val stackPointer = Register("R31", 64)
+
+  // this is the mapping from offsets/positions on the stack to their representative DS nodes
+  val stackMapping: mutable.Map[BigInt, DSN] =
+    computeDomain(IntraProcIRCursor, Set(proc)).toSeq.sortBy(_.toShortString).foldLeft(Map[BigInt, DSN]()) {
+      (results, pos) => stackBuilder(pos, results)
+    }.to(collection.mutable.Map)
+
+
+  /**
+   * this function takes a stackMapping and updates it based on a memory load and memory store
+   * @param pos memory load or store IL position
+   * @param index memory location of load or store
+   * @param size size of the load or store
+   * @param m stack mapping
+   * @return updated stack mapping
+   */
+  private def visitStackAccess(pos: CFGPosition, index: Expr, size: Int, m: Map[BigInt, DSN]) : Map[BigInt, DSN] =
+    val byteSize = (size.toDouble / 8).ceil.toInt
+    index match
+      case BinaryExpr(op, arg1: Variable, arg2) if varToSym.contains(pos) && varToSym(pos).contains(arg1) &&
+        evaluateExpression(arg2, constProp(pos)).isDefined =>
+        var offset = evaluateExpression(arg2, constProp(pos)).get.value
+        varToSym(pos)(arg1).foldLeft(m) { // go through all the symbolic accesses tied to arg1 at pos
+          (m, sym) =>
+            sym match
+              case SymbolicAccess(accessor, StackLocation(regionIdentifier, proc, size), symOffset) => // only consider stack accesses
+                offset = offset + symOffset
+                createStackMapping(pos.toShortString, offset, m, byteSize)
```

Comment:
`offset` being a var here seems incorrect - if varToSym(pos)(arg1) contained multiple stack accesses (which it doesn't seem to in practice but I can't be sure about that), then any subsequent offsets would be incorrect as they would not relate to the base offset?

Reply:
You are correct. I think there is an implicit assumption here from the assertion in local that there is only on stack location associated with each variable at each cfgposition. However, that assertions doesn't always hold.